### PR TITLE
Fix(ticket#165): Reordering Users According to Shift  and Timings 

### PR DESCRIPTION
**Pull request by nadirhussain786**
## Change Request

This PR updates the member list to organize members according to their assigned shifts. Once the shift updates are applied, the members will be automatically sorted in descending order based on their user creation time.

### Resources

image

### CODE REVIEW

File: src/hooks/useGroupedData.ts [hunk 1]
```diff
@@ -9,19 +9,34 @@ function useGroupedData<T extends Record<string, unknown>>(
       return [];
     }
 
-    const groupedData = data?.reduce((acc: Record<string, T[]>, item: T) => {
+    const groupedData = data.reduce((acc: Record<string, T[]>, item: T) => {
       const groupKey = item[key];
       if (!acc[groupKey as string]) {
```
Comment (Monfernape) [inline src/hooks/useGroupedData.ts:14]: No type assertions please.

File: src/hooks/useGroupedData.ts [hunk 2]
```diff
@@ -9,19 +9,34 @@ function useGroupedData<T extends Record<string, unknown>>(
...
       acc[groupKey as string].push(item);
       return acc;
     }, {});
-    return Object?.entries(groupedData)?.map(([shift, row]) => ({
-      shift,
-      row,
-    }));
+
+    const sortedKeys = Object.keys(groupedData).sort((a, b) => {
+      const order = ["A", "B", "C" , "D"];
```
Comment (Monfernape) [inline src/hooks/useGroupedData.ts:22]: This should be using enums

File: src/hooks/useGroupedData.ts [hunk 3]
```diff
@@ -9,19 +9,34 @@ function useGroupedData<T extends Record<string, unknown>>(
...
+      return order.indexOf(a) - order.indexOf(b);
+    });
+
+    return sortedKeys.map((shift) => {
+      const sortedMembers = groupedData[shift].sort((a, b) => {
+        const dateA = new Date(a.created_at as string).getTime();
```
Comment (Monfernape) [inline src/hooks/useGroupedData.ts:28]: I don't think we'd need type assertion here. Isn't created_at known to be string?

File: src/hooks/useGroupedData.ts [hunk 4]
```diff
@@ -9,19 +9,34 @@ function useGroupedData<T extends Record<string, unknown>>(
       return [];
     }
 
-    const groupedData = data?.reduce((acc: Record<string, T[]>, item: T) => {
+    const groupedData = data.reduce((acc: Record<string, T[]>, item: T) => {
```
Comment (muhammad-bilal-ali-khokhar) [inline src/hooks/useGroupedData.ts:12]: @nadirhussain786 i think data possibly undefine

File: src/hooks/useGroupedData.ts (hunk 1 above)
Comment (nadirhussain786) [reply src/hooks/useGroupedData.ts]: Sir @Monfernape , No need for this hook anymore because I've fixed it local file.

**Comment by nadirhussain786** [comment]
Sir, No need for this hook anymore because I've fixed it local file.

File: src/app/members/components/MemberList.tsx [hunk 5]
```diff
@@ -18,12 +18,12 @@ import {
 } from "@/components/ui/table";
 import { Plus as PlusIcon } from "lucide-react";
 import { useRouter } from "next/navigation";
-import useGroupedData from "@/hooks/useGroupedData";
 import { useToast } from "@/hooks/use-toast";
 import { Member } from "@/types";
 import { UserStatusBadge } from "@/common/StatusBadge/UserStatusBadge";
 import { StandardPage } from "@/common/StandardPage/StandardPage";
 import { Routes } from "@/lib/routes";
+import { Shift } from "@/constant/constant";
```
Comment (Monfernape) [inline src/app/members/components/MemberList.tsx:26]: This file name is weird.

File: src/app/members/components/MemberList.tsx [hunk 6]
```diff
@@ -93,7 +93,15 @@ export const MemberList = ({ members }: { members: Props }) => {
     }
   }, [success, toast]);
 
-  const groupedData = useGroupedData(data, "shift");
+  const allShifts = [Shift.ShiftA, Shift.ShiftB, Shift.ShiftC , Shift.ShiftD];
+
+  const groupedData = allShifts.map((shift) => {
+    const usersInShift = data.filter((user) => user.shift === shift);
+    return {
+      shift: shift,
```
Comment (Monfernape) [inline src/app/members/components/MemberList.tsx:101]:
```suggestion
      shift,
```

In JS, when key and values are same, we can omit values.

**Review by Monfernape — APPROVED** [review]

File: src/hooks/useGroupedData.ts (hunk 3 above)
Comment (nadirhussain786) [reply src/hooks/useGroupedData.ts]: Fixed

File: src/hooks/useGroupedData.ts (hunk 4 above)
Comment (nadirhussain786) [reply src/hooks/useGroupedData.ts]: Fixed

**Pull request merged**